miku-server/src/main/scala/org/http4s/miku/MikuBuilder.scala

```diff
@@ -0,0 +1,492 @@
+        getContext() match {
+          case Some((ctx, auth)) => //Ignore client authentication for now
+            val engine = ctx.createSSLEngine()
+            engine.setUseClientMode(false)
+            auth match {
+              case NoClientAuth => ()
+              case ClientAuthRequired =>
+                engine.setNeedClientAuth(true)
+              case ClientAuthOptional =>
+                engine.setWantClientAuth(true)
+            }
+            pipeline.addLast("ssl", new SslHandler(engine))
+
+          case None => //Do nothing
+        }
+
+        // Netty HTTP decoders/encoders/etc
+        pipeline.addLast(
+          "decoder",
+          new HttpRequestDecoder(maxInitialLineLength, maxHeaderSize, maxChunkSize))
+        pipeline.addLast("encoder", new HttpResponseEncoder())
+        pipeline.addLast("decompressor", new HttpContentDecompressor())
+
+        idleTimeout match {
+          case Duration.Inf => //Do nothing
+          case Duration(timeout, timeUnit) =>
+            pipeline.addLast("idle-handler", new IdleStateHandler(0, 0, timeout, timeUnit))
+
+        }
+
+        val requestHandler = newRequestHandler()
+
+        // Use the streams handler to close off the connection.
+        pipeline.addLast(
+          "http-handler",
+          new HttpStreamsServerHandler(Seq[ChannelHandler](requestHandler).asJava))
+
+        pipeline.addLast("request-handler", requestHandler)
+
+        // And finally, register the channel with the event loop
+        val childChannelEventLoop = eventLoop.next()
+        childChannelEventLoop.register(connChannel)
+        allChannels.add(connChannel)
+        ()
+      }
+    }
+  }
+
+  private def addChannelOptions(b: Bootstrap) =
+    nettyChannelOptions.foldLeft(b)((boot, o) => boot.option(o._1, o._2))
+
+  private def bootstrapNative(
+      serverChannelEventLoop: EventLoopGroup,
+      channelPublisher: HandlerPublisher[Channel],
+      address: InetSocketAddress): Bootstrap =
+    addChannelOptions(
+      new Bootstrap()
+        .channel(classOf[EpollServerSocketChannel])
+        .group(serverChannelEventLoop))
+      .option(ChannelOption.AUTO_READ, java.lang.Boolean.FALSE) // publisher does ctx.read()
+      .handler(channelPublisher)
+      .localAddress(address)
+
+  private def bootstrapNIO(
+      serverChannelEventLoop: EventLoopGroup,
+      channelPublisher: HandlerPublisher[Channel],
+      address: InetSocketAddress): Bootstrap =
+    addChannelOptions(
+      new Bootstrap()
+        .channel(classOf[NioServerSocketChannel])
+        .group(serverChannelEventLoop))
+      .option(ChannelOption.AUTO_READ, java.lang.Boolean.FALSE) // publisher does ctx.read()
+      .handler(channelPublisher)
+      .localAddress(address)
+
+  protected def getBootstrap(
+      serverChannelEventLoop: EventLoopGroup,
+      channelPublisher: HandlerPublisher[Channel],
+      address: InetSocketAddress): Bootstrap = transport match {
```

Comment:
This logic looks tangled with the transport logic in `start`.  Is there a good way to do this transport sniffing once?  Maybe it doesn't really matter.

Reply:
I can look to reorder the code for it, sure. I wrote that at 4am